subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/MultipleCandidateMatcher.java

```diff
@@ -345,6 +362,8 @@ private void disambiguateWithExtraAttributes() {
      * from an artifact repository. We always check whether the schema has a more strongly
      * typed version of an attribute and use that one instead to apply its disambiguation
      * rules.
+     *
+     * Must be called before attempting to call {@code getAttribute} with {@code a} &gt; {@code requestedAttributes.size()}.
```

Comment:
Is there any better way we could enforce this temporal coupling somehow, rather then just documenting it?

Reply:
Yes, a refactor. I've since started taking a look at this and it seems possible, but its definitely out of scope for this PR. A lot of the code in this class relies on this temporal coupling, so resolving it would take more changes in this PR than I really want. 

I'm happy to revisit this with a follow-up PR though

Reply:
Follow up PR regarding this: https://github.com/gradle/gradle/pull/21723